play/plugin/src/test/kotlin/com/github/triplet/gradle/play/tasks/GenerateResourcesIntegrationTest.kt

```diff
@@ -380,6 +381,7 @@ class GenerateResourcesIntegrationTest : IntegrationTestBase(), SharedIntegratio
     }
 
     @Test
+    @Ignore("Previously this test worked because empty catalogs were generated and passed into the task")
```

Comment:
To be verified if this test/implementation for it should be rewritten

Reply:
The check for this is here: https://github.com/Triple-T/gradle-play-publisher/blob/4273731c8126cdbf071300882718ec2ae12f3e3c/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/GenerateResources.kt#L102-L107

This test absolutely has to pass because we rely on it for correctness. The way we pick which resource to use for a particular flavor is dependent on the parent of the `play` directory being a valid flavor name: https://github.com/Triple-T/gradle-play-publisher/blob/4273731c8126cdbf071300882718ec2ae12f3e3c/play/plugin/src/main/kotlin/com/github/triplet/gradle/play/tasks/GenerateResources.kt#L449

Anyway, off the top of my head I think you can fix `name != PLAY_PATH` by combining two climbUpTos and if both of them are non-null then you have two play directories. There might be a cleaner way though.

Reply:
@SUPERCILEX could you please elaborate more on what this test is supposed to verify? 🙏 

Looking at those two it seems like those test are identical except for different flavor name - it says "`play` file name is not allowed", but I can't grasp my head around it - it sounds like only "play" flavor is allowed and others are not (which is unexpected I guess)?
```kotlin
    @Test
    fun `Flavor named 'play' is allowed`() {
        // language=gradle
        val config = """
            flavorDimensions 'pricing'
            productFlavors {
                play { dimension 'pricing' }
            }
        """.withAndroidBlock()

        val result = execute(config, "generatePlayReleasePlayResources")

        result.requireTask(taskName("PlayRelease"), outcome = SUCCESS)
    }

    @Test
    @Ignore("Previously this test worked because empty catalogs were generated and passed into the task")
    fun `'play' file name is not allowed`() {
        // language=gradle
        val config = """
            flavorDimensions 'pricing'
            productFlavors {
                illegalPlay { dimension 'pricing' }
            }
        """.withAndroidBlock()

        val result = executeExpectingFailure(
                config, "generateIllegalPlayReleasePlayResources")

        result.requireTask(taskName("IllegalPlayRelease"), outcome = FAILED)
    }
```

Please read my PR's description on how now parameters provided to
```kotlin
abstract class Validator : WorkAction<Validator.Params> {
        override fun execute() {
            for (file in parameters.files.get()) {
                file.validate()
            }
        }
```
have changed with the recent version.

I'm happy to improve it but would really appreciate your explanation in here 😄

Reply:
Shoot, forgot about that, makes life more complicated. For understanding the tests, it's easiest to look at the fixtures: https://github.com/Triple-T/gradle-play-publisher/tree/master/play/plugin/src/test/fixtures/GenerateResourcesIntegrationTest/src. Every directory in there is a flavor, so you'll find the play and illegalPlay dirs in there.

I've looked at this some more and I think we might be able to fix this by replacing `it.asFile == this` with `it.asFile == climbUpTo(PLAY_PATH)`. That check is looking for file names that are `play`, but not part of the root input dirs. So if we look through every input file and find its parent play dir, it should always be a root play dir (I think).

Reply:
@SUPERCILEX please take a look at dbcd5a2.

My confusion came from this test's name being a little bit misleading I think.

From what I understand now is that those two cases are illegal:
- One is not allowed to have any files named "play"
`project/my-project/src/flavorName/play/**/play.txt`

- One is not allowed to have any child directories named play - only root directory can have "play" name:
`project/my-project/src/flavorName/play/**/play/*.*`

Even if the flavor is named `play`, this would be the root directory: `src/play/play` - my changes made it look only at  the child directories of the subdirectory.
So `src/play/play/default-listing.txt` is ok, `src/play/play/**/play/file.txt` is not. 

I tested this change with both old Gradle version and new one and in both cases it now passes.

Reply:
> One is not allowed to have any files named "play"
project/my-project/src/flavorName/play/**/play.txt

Yes to the sentence, no to the example. `/play.txt` should be fine, `/play` isn't. On that note, can you add two test cases for this: one allowing `play.foo` and one disallowing plain `play` files? Then check that the tests pass pre-PR and we'll be in the clear for behavior changes.

---

I like the simplicity of your approach, but I'm a little bothered by the performance implications. Previously, we only did validation on changed files (since the input files are incremental). Now, we always run a full directory expansion. I think just because of that'd I'd still prefer the "climb up every file's path and see if its root play dir is contained in the input dir set" approach.

Reply:
Thank you - I was so focused on 'file' that I forgot that you could actually have a file `play` without an extension.

Got actually a little bit misled because `GenerateResourcesIntegrationTest` fixture does not contain a file named "play" but only a directory named "play" (listings/en-US/play/.keep).
This is the reason why the test would start to fail, as there was no file "play" in the input anymore with newer Gradle version (previously there was a a list of files and also directories, which were then treated `asFile`). ".keep" got ignored.
```
src/main/play/default-language.txt
src/main/play/listings/de-DE/full-description.txt
src/main/play/listings/en-US/full-description.txt
src/main/play/listings/en-US/graphics/phone-screenshots/foo.jpg
src/main/play/listings/en-US/graphics/tablet-screenshots/baz.jpg
src/main/play/listings/en-US/short-description.txt
src/main/play/listings/en-US/title.txt
src/main/play/listings/fr-FR/full-description.txt
src/main/play/listings/fr-FR/graphics/phone-screenshots/bar.jpg
src/main/play/listings/fr-FR/short-description.txt
src/main/play/listings/fr-FR/title.txt
src/main/play/products/sku.json
src/main/play/release-notes/en-US/default.txt
src/main/play/release-notes/fr-FR/default.txt
```

> One is not allowed to have any child directories named play - only root directory can have "play" name:

Is this also valid?

Reply:
> Is this also valid?

Sorry, which part? Or do you mean in general? If so, then yeah your understanding is correct.